src/lib.rs

```diff
@@ -352,75 +366,25 @@ pub trait PolynomialCommitment<F: Field>: Sized {
         vk: &Self::VerifierKey,
-        commitments: impl IntoIterator<Item = &'a LabeledCommitment<Self::Commitment>>,
+        commitments: impl Clone + IntoIterator<Item = &'a LabeledCommitment<Self::Commitment>>,
         query_set: &QuerySet<F>,
         evaluations: &Evaluations<F>,
         proof: &Self::BatchProof,
         opening_challenges: &dyn Fn(u64) -> F,
         rng: &mut R,
     ) -> Result<bool, Self::Error>
     where
-        Self::Commitment: 'a,
-    {
-        let commitments: BTreeMap<_, _> = commitments.into_iter().map(|c| (c.label(), c)).collect();
-        let mut query_to_labels_map = BTreeMap::new();
-        for (label, (point_label, point)) in query_set.iter() {
-            let labels = query_to_labels_map
-                .entry(point_label)
-                .or_insert((point, BTreeSet::new()));
-            labels.1.insert(label);
-        }
-
-        // Implicit assumption: proofs are order in same manner as queries in
-        // `query_to_labels_map`.
-        let proofs: Vec<_> = proof.clone().into();
-        assert_eq!(proofs.len(), query_to_labels_map.len());
-
-        let mut result = true;
-        for ((_point_label, (point, labels)), proof) in query_to_labels_map.into_iter().zip(proofs)
-        {
-            let mut comms: Vec<&'_ LabeledCommitment<_>> = Vec::new();
-            let mut values = Vec::new();
-            for label in labels.into_iter() {
-                let commitment = commitments.get(label).ok_or(Error::MissingPolynomial {
-                    label: label.to_string(),
-                })?;
-
-                let v_i =
-                    evaluations
-                        .get(&(label.clone(), *point))
-                        .ok_or(Error::MissingEvaluation {
-                            label: label.to_string(),
-                        })?;
-
-                comms.push(commitment);
-                values.push(*v_i);
-            }
-
-            let proof_time = start_timer!(|| "Checking per-query proof");
-            result &= Self::check_individual_opening_challenges(
-                vk,
-                comms,
-                *point,
-                values,
-                &proof,
-                opening_challenges,
-                Some(rng),
-            )?;
-            end_timer!(proof_time);
-        }
-        Ok(result)
-    }
-
+        Self::Commitment: 'a;
+    
     /// open_combinations but with individual challenges
     fn open_combinations_individual_opening_challenges<'a>(
         ck: &Self::CommitterKey,
         linear_combinations: impl IntoIterator<Item = &'a LinearCombination<F>>,
-        polynomials: impl IntoIterator<Item = &'a LabeledPolynomial<F>>,
-        commitments: impl IntoIterator<Item = &'a LabeledCommitment<Self::Commitment>>,
+        polynomials: impl Clone + IntoIterator<Item = &'a LabeledPolynomial<F>>,
+        commitments: impl Clone + IntoIterator<Item = &'a LabeledCommitment<Self::Commitment>>,
```

Comment:
Not related to the changes of the pull request, but to the ``opening_challenges()`` functions when using it for linear combinations.
Beside a secure initialization, the caller is responsible to define ``opening_challenges()`` correctly, combining random challenges with the LC coefficients in a proper manner. This seems to me a bit of an overload of the natural meaning of challenges.